metaserver/src/main/scala/scala/meta/languageserver/ScalametaServices.scala

```diff
@@ -246,6 +259,19 @@ class ScalametaServices(
       sourceChangeSubscriber.onNext(input)
       ()
     }
+    .notification(td.willSave) { _ =>
+      ()
+    }
+    .requestAsync(td.willSaveWaitUntil) { params =>
```

Comment:
I'm not sure I'm doing right thing here. Probably it should be just `.request`? Why does `documentFormattingProvider.format` return response errors explicitly?

Reply:
I have a branch somewhere locally where I refactored the lsp4s API for requests to return `T` instead of `Either[Response.Error, T]`. Errors can instead be thrown as exception. I think this refactoring cleaned up a lot of code.

The reason it's `requestAsync` is simply because we don't have a combinator for synchronous explicit error handling.

Reply:
OK, I see

Reply:
Btw, do you mean this branch: https://github.com/olafurpg/language-server/tree/refactoring? You mentioned it in #174.

Reply:
correct, I never polished it into a PR because there is too much happening in that PR. The relevant commit is https://github.com/olafurpg/language-server/commit/3a8c1a7a797a8fdea80cd913aa8f068de3624621